src/python/pants/engine/target.py

```diff
@@ -2687,6 +2688,25 @@ class ValidatedDependencies:
     pass
 
 
+@dataclass(frozen=True)
+class DependenciesRuleActionRequest:
+    """A request to return the applicable dependency rule action for each dependency of a target."""
+
+    address: Address
+    dependencies: Addresses
+    description_of_origin: str = dataclasses.field(hash=False, compare=False)
+
+
+@dataclass(frozen=True)
+class DependenciesRuleAction:
+    address: Address
+    dependencies_rule: FrozenDict[Address, DependencyRuleAction]
+
+    def execute_actions(self) -> None:
+        for dependency, action in self.dependencies_rule.items():
+            action.execute(description_of_origin=f"{self.address} on {dependency}")
```

Comment:
Each address has a rule--a `DependencyRuleAction`--but a collection of dependencies does not share one action, so I think actions should be plural. Maybe do this?

```suggestion
class DependenciesRuleActions:
    address: Address
    dependencies_rules: FrozenDict[Address, DependencyRuleAction]

    def execute_actions(self) -> None:
        for dependency, action in self.dependencies_rules.items():
            action.execute(description_of_origin=f"{self.address} on {dependency}")
```

`DependenciesRulesAction` also has the same shortcoming in that it suggests there is only one action for the collection for rules. So, I would go for either `DependenciesRulesActions` or `DependenciesRuleActions`.

Reply:
You could also call this `DependenciesRules` where a "rule" consists of an address and an action.

Reply:
I totally see your point, and could be persuaded to agree, after presenting my train of thought for what I have here ;) 

> dependencies does not share one action

True, but each dependency have only a single action, not many, so it is the `RuleAction` for each dependency. So the name of the class does not convey whether the rule action is shared or not, but looking at the field type for `dependencies_rule` does.
Where as I find that:
```python
dependencies_rules: FrozenDict[Address, DependencyRuleAction]
```
suggests a typo, there I would've expected something like this to match the name:
```python
dependencies_rules: FrozenDict[Address, Sequence[DependencyRuleAction]]
```

Reply:
I see where that shifts the confusion. Naming is hard.

Ok. Well leave it as is then. I don't have any suggestions I like better (I played with adding `map` to the name, but that's not great either). /shrug

Reply:
agree, naming is a hard nut.